File: backend/app/modules/client/api/file.py
"""
文件上传接口
按场景目录管理上传文件（用户头像、运单照片等）
"""

import os
import uuid
from pathlib import Path

from fastapi import APIRouter, Depends, UploadFile, File, Form
from starlette.requests import Request

from app.core.dependencies import get_current_user
from app.core.security import TokenData
from app.common.response import success
from app.common.exceptions import BizException

router = APIRouter()

ALLOWED_IMAGE_EXTS = {".jpg", ".jpeg", ".png", ".gif", ".webp", ".bmp"}
MAX_FILE_SIZE = 5 * 1024 * 1024  # 5MB

UPLOAD_ROOT = Path(__file__).resolve().parents[4] / "uploads"

ALLOWED_SCENES = {"avatar", "waybill", "vehicle", "document"}


@router.post("/upload")
async def upload_file(
    file: UploadFile = File(...),
    scene: str = Form(default="avatar"),
    current_user: TokenData = Depends(get_current_user),
):
    """
    上传文件到指定场景目录
    - scene: 场景标识，决定存储子目录（avatar/waybill/vehicle/document）
    - 返回文件的相对访问路径
    """
    if scene not in ALLOWED_SCENES:
        raise BizException(f"不支持的上传场景: {scene}")

    original_name = file.filename or "unknown"
    ext = os.path.splitext(original_name)[1].lower()
    if ext not in ALLOWED_IMAGE_EXTS:
        raise BizException(f"不支持的文件类型: {ext}")

    content = await file.read()
    if len(content) > MAX_FILE_SIZE:
        raise BizException("文件大小不能超过5MB")

    scene_dir = UPLOAD_ROOT / scene
    scene_dir.mkdir(parents=True, exist_ok=True)

    new_filename = f"{uuid.uuid4().hex}{ext}"
    file_path = scene_dir / new_filename
    file_path.write_bytes(content)

    relative_url = f"/uploads/{scene}/{new_filename}"

    return success(data={
        "url": relative_url,
        "name": original_name,
    })

File: frontend/driver-h5/src/api/file.ts
import request from './request';

export interface UploadResult {
  url: string;
  name: string;
}

/**
 * 上传单张图片到司机端文件服务。
 * scene 见后端 ALLOWED_SCENES：task_loading（装卸车）/ task_receipt（回单）/ avatar 等。
 */
export async function uploadImage(
  file: File,
  scene = 'task_receipt'
): Promise<UploadResult> {
  const formData = new FormData();
  formData.append('file', file);
  formData.append('scene', scene);
  const res = await request.post<{ code: number; message?: string; data?: UploadResult }>(
    '/file/upload',
    formData,
    { headers: { 'Content-Type': 'multipart/form-data' } }
  );
  const body = res.data;
  if (body?.data?.url) {
    return body.data;
  }
  return Promise.reject(new Error(body?.message || '上传失败'));
}
